aten/src/ATen/native/cuda/Sorting.cu

```diff
@@ -0,0 +1,451 @@
+      });
+  return std::forward_as_tuple(values, indices);
+}
+
+std::tuple<Tensor&, Tensor&> median_with_indices_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim,
+    bool ignore_nan) {
+  NoNamesGuard guard;
+
+  dim = at::maybe_wrap_dim(dim, self.dim());
+  Tensor in = self.dim() > 0 ? self : self.unsqueeze(0);
+
+  int64_t size = in.size(dim);
+  TORCH_CHECK(
+      size > 0,
+      "median() operation does not have an identity for empty input tensor");
+
+  checkDeviceType("median", {values, indices}, self.device().type());
+  checkScalarType("median", {indices, "indices", 1}, kLong);
+  checkSameType("median", {values, "values", 0}, {self, "self", 2});
+
+  TORCH_CHECK(
+      self.dim() <= MAX_TENSORINFO_DIMS,
+      "median() cannot operate on more than ",
+      MAX_TENSORINFO_DIMS,
+      " dimensions");
+
+  std::vector<int64_t> out_shape = self.sizes().vec();
+  if (self.dim() > 0) {
+    if (keepdim) {
+      out_shape[dim] = 1;
+    } else {
+      out_shape.erase(out_shape.begin() + dim);
+    }
+  }
+
+  values.resize_(out_shape);
+  indices.resize_(out_shape);
+
+  // Only launch kernel for non-empty tensors
+  if (self.numel() > 0) {
+    // Ensure #dim is the same for all tensors required for reduction
+    Tensor vals = keepdim && self.dim() > 0 ? values : values.unsqueeze(dim);
+    Tensor inds = keepdim && self.dim() > 0 ? indices : indices.unsqueeze(dim);
+
+    AT_DISPATCH_ALL_TYPES_AND(
+        at::ScalarType::Half, self.scalar_type(), "median_out_impl", [&] {
+          if (cuda::detail::canUse32BitIndexMath(vals) &&
+              cuda::detail::canUse32BitIndexMath(inds) &&
+              cuda::detail::canUse32BitIndexMath(in)) {
+            run_launcher<scalar_t, uint32_t>(
+                vals, inds, in, dim, MedianLauncher(ignore_nan));
+          } else {
+            run_launcher<scalar_t, uint64_t>(
+                vals, inds, in, dim, MedianLauncher(ignore_nan));
+          }
+        });
+
+    AT_CUDA_CHECK(cudaGetLastError());
+  }
+
+  guard.reset();
+  namedinference::propagate_names_for_reduction(values, self, dim, keepdim);
+  namedinference::propagate_names_for_reduction(indices, self, dim, keepdim);
+
+  return std::forward_as_tuple(values, indices);
+}
+
+Tensor median_impl(const Tensor& self, bool ignore_nan) {
+  NoNamesGuard guard;
+
+  int64_t size = self.numel();
+  TORCH_CHECK(size > 0, "median() input tensor cannot be empty");
+
+  // Sort input tensor to efficiently query for median element
+  Tensor sorted = std::get<0>(self.flatten().sort());
+
+  if (!ignore_nan) {
+    // For torch.median return either the middle element or nan (sorted as
+    // largest) if there are any
+    int64_t k = (size - 1) / 2;
+    return at::where(sorted[-1].isnan(), sorted[-1], sorted[k]);
+  } else {
+    // For torch.nanmedian return the middle element among the non-nan values
+    Tensor k = ((size - 1) - sorted.isnan().sum()) / 2;
```

Comment:
Instead of creating a tensor here would it be better to use `item()` on the result of `sum`?